keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,1000 @@
+than using a global logger instance. The overhead will be measured for
+different scenarios. If a function uses a logger repeatedly, it should retrieve
+the logger once and then use that instance.
+
+More expensive than logger lookup are `WithName` and `WithValues` and creating
+a new context with the modified logger. This needs to be used cautiously in
+performance-sensitive code. A possible compromise is to enhance logging with
+such additional information only at higher log levels.
+
+#### Transition
+
+Once some components start to rely on `k8s.io/klogr` exclusively instead of
+klog, users of those components must remember to initialize klog *and* the
+default logger in `k8s.io/klogr`. Features provided only by klog like writing
+to files become unavailable.
+
+This breaking change can be postponed until after the [deprecation of klog
+features](https://github.com/kubernetes/enhancements/issues/2845) is complete
+by using a logger which passes log messages through to klog. That allows
+converting components already now. Then later, the klog dependency in
+`k8s.io/klogr` can be replaced with a stand-alone logger which produces the same
+text output.
+
+#### Pitfalls during usage
+
+Code reviews must catch the following incorrect patterns.
+
+##### Logger in context and function not consistent
+
+Incorrect:
+
+```go
+func foo(ctx context.Context) {
+   logger := klogr.FromContext(ctx)
+   ctx = klogr.NewContext(ctx, logger.WithName("foo"))
+   doSomething(ctx)
+   // BUG: does not include WithName("foo")
+   logger.Info("Done")
+}
+```
+
+Correct:
+
+```go
+func foo(ctx context.Context) {
+   logger := klogr.FromContext(ctx).WithName("foo")
+   ctx = klogr.NewContext(ctx, logger)
+   doSomething(ctx)
+   logger.Info("Done")
+}
+```
+
+In general, manipulating a logger and the corresponding context should be done
+in separate lines.
+
+##### Overwriting context not needed initially, forgotten later
+
+Initial, correct code with contextual logging:
+
+```go
+func foo(ctx context.Context) {
+   logger := klogr.FromContext(ctx).WithName("foo")
+   doSomething(logger)
+   logger.Info("Done")
+}
+```
+
+A line with `ctx = klogr.NewContext(ctx, logger)` could be added above (it
+compiles), but it causes unnecessary overhead and some linters complain about
+"new value not used".
+
+However, care then must be taken when adding code later which uses `ctx`:
+
+```go
+func foo(ctx context.Context) {
+   logger := klogr.FromContext(ctx).WithName("foo")
+   doSomething(logger)
+   // BUG: ctx does not contain the modified logger
+   doSomethingWithContext(ctx)
+   logger.Info("Done")
+}
+```
+
+##### Redundant key/value pairs
+
+When the caller already adds a certain key to the logger, the callee should not
+add it again because it is redundant. It makes log output larger (for
+performance reasons, loggers like zap and klog do not filter out duplicates)
+and causes additional overhead.
```

Comment:
to comply with this guidance, doesn't that mean callers will very frequently have to speculatively create nested loggers and set `WithValue` repeatedly for many of the parameters they are going to pass to subcalls, or risk logging by the subcalls omitting significant information? what is the allocation cost of creating new loggers and populating values like that approaching ~1:1 with function calls? currently, we generally avoid allocations if log lines are excluded by verbosity settings, but this seems like it would incur those costs regardless of verbosity settings

Reply:
The guidance is not that callers should add common values, it's the other way around: if it is known (for example, from documentation about how a plugin for the scheduler is invoked) that a certain parameter always gets logged via `WithValue`, then the code was meant to not add it again.

Whether the caller decides that a certain value is important enough to be added to all log entries is a separate question and needs careful evaluation of the benefit vs. performance overhead.

But I agree that it is hard to formulate this as a generally applicable recommendation and, more importantly, with the addition of the feature gate it also cannot be relied upon anymore: if the feature gate is off, the logger will not have any additional values, so individual log calls must still add it.

To make that work when the feature gate is on, I added key/value deduplication for text output. For JSON we end up with some duplication that the log consumer will resolve when parsing the log line. Given that the WithValues feature needs to be use very selectively I think that's okay.

I've updated this section to reflect this new approach, i.e. the recommendation now is the opposite with an explanation why.